docs/design/upgrade.md

```diff
@@ -0,0 +1,182 @@
+  apply(input: string): string;
+
+  // Returns the order this change should be applied. Normally the line number.
+  // Changes are applied from the bottom of a file.
+  order(): number;
+
+  // The description of this change. This will be outputted in a dry run.
+  toString(): string;
+}
+```
+
+Some available example of a change implementation:
+
+```typescript
+// Will add text to the source code.
+class AppendChange implements Change {
+  constructor(line: number, column: number, toAdd: string);
+}
+// Will remove text from the source code.
+class RemoveChange implements Change {
+  constructor(line: number, column: number, toRemove: string);
+}
+// Will replace text from the source code.
+class ReplaceChange implements Change {
+  constructor(line: number, column: number, oldText: string, newText: string);
+}
+// Will output a message for the user to fulfill.
+class MessageChange implements Change {
+  constructor(text: string);
+}
+```
+
+Some higher function changes are provided to allow an easier way to create changes that are harder to make:
+
+```typescript
+// Rename a class to another name. Need to pass a symbol as input, as just
+// a string is not enough.
+declare function renameClass(project: Angular2Project,
+                             from: ts.Symbol, to: string): Observable<Change>;
+```
+
+Some higher level changes examples:
+
+```typescript
+// Add a new Import if the import isn't already in the file.
+class AddImportChange implements Change {
+  constructor(name: string);
+}
+```
+
+# Upgrade Command
+A new command must be added to the CLI;
+
+    ng upgrade [options]
+    Upgrade the project to a newer version.
+
+    Options:
+      --package=<name>  The package to upgrade. By default, upgrade every packages
+                        that are outdated.
+      --to=<semver>     Move to a specific version of Angular. Only valid with
+                        --package and --dry-run.
+                        Default: latest.
+      --from=<semver>   Specify the version the project is currently using. Only
+                        valid with --package and --dry-run.
+                        Default: the installed version is used.
+      --commit=<bool>   Whether the tool should automatically create a commit.
+                        Default: false.
+      --message=<msg>   The commit message to use. Only valid with --commit.
+      --dry-run         Only output the commands and refactor steps that would be run,
+                        but do nothing.
+                        Default: false.
+
+## Process
+The command will execute the following tasks in order:
+
+1. If the `angular-cli` is not the latest version, return with an error message suggesting an upgrade. This step is necessary even if the version is supported as a upgrade, because updates to the upgrade steps might have occurred.
+1. If the project is in a git repository, verify that the git is currently on the latest origin/master. If not, exit early.
```

Comment:
If a project *is* a git repository I think the check should be that there are no changes pending (staged or unstaged). Not that the user is on a given branch. I may want to upgrade a previous release (not latest on master) for a number of different business cases. If anything I think a warning could be shown, but definitely not a hard stop.